File: tests/python/user_repository.py
"""
Repository for managing User data persistence.
"""
from typing import Optional, List
from user import User


class UserRepository:
    """
    Repository for managing User data persistence.
    Uses in-memory storage for demonstration purposes.
    """

    def __init__(self):
        self._database: dict[int, User] = {}
        self._id_counter: int = 1

    def save(self, user: User) -> User:
        """
        Save a user to the database.
        If the user has an ID, it updates; otherwise creates a new user.
        """
        if user.id is None:
            user.id = self._id_counter
            self._id_counter += 1

        self._database[user.id] = user
        return user

    def find_by_id(self, user_id: int) -> Optional[User]:
        """Find a user by their ID."""
        return self._database.get(user_id)

    def find_by_username(self, username: str) -> Optional[User]:
        """Find a user by their username."""
        for user in self._database.values():
            if user.username == username:
                return user
        return None

    def find_all_active(self) -> List[User]:
        """Find all active users."""
        return [user for user in self._database.values() if user.active]

    def delete_by_id(self, user_id: int) -> bool:
        """Delete a user by ID."""
        if user_id in self._database:
            del self._database[user_id]
            return True
        return False

    def count(self) -> int:
        """Get count of all users."""
        return len(self._database)

File: tests/java/UserRepository.java
package com.example.demo.repository;

import com.example.demo.model.User;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Repository for managing User data persistence.
 * Uses in-memory storage for demonstration purposes.
 */
public class UserRepository {
    private final Map<Long, User> database = new ConcurrentHashMap<>();
    private final AtomicLong idGenerator = new AtomicLong(1);

    /**
     * Save a user to the database.
     * If the user has an ID, it updates; otherwise creates a new user.
     */
    public User save(User user) {
        if (user.getId() == null) {
            user.setId(idGenerator.getAndIncrement());
        }
        database.put(user.getId(), user);
        return user;
    }

    /**
     * Find a user by their ID.
     */
    public Optional<User> findById(Long id) {
        return Optional.ofNullable(database.get(id));
    }

    /**
     * Find a user by their username.
     */
    public Optional<User> findByUsername(String username) {
        return database.values().stream()
                .filter(user -> user.getUsername().equals(username))
                .findFirst();
    }

    /**
     * Find all active users.
     */
    public List<User> findAllActive() {
        return database.values().stream()
                .filter(User::isActive)
                .toList();
    }

    /**
     * Delete a user by ID.
     */
    public boolean deleteById(Long id) {
        return database.remove(id) != null;
    }

    /**
     * Get count of all users.
     */
    public long count() {
        return database.size();
    }
}
